modules/currencies/src/lib.rs

```diff
@@ -637,12 +657,332 @@ impl<T: Config> MultiReservableCurrency<T::AccountId> for Pallet<T> {
+			CurrencyId::Erc20(_) => {
+				if amount.is_zero() {
+					return WithdrawConsequence::Success;
+				}
+
+				if <Self as fungibles::Inspect<_>>::total_issuance(asset_id)
+					.checked_sub(&amount)
+					.is_none()
+				{
+					return WithdrawConsequence::Underflow;
+				}
+
+				if <Self as fungibles::Inspect<_>>::balance(asset_id, who)
+					.checked_sub(&amount)
+					.is_none()
+				{
+					return WithdrawConsequence::NoFunds;
+				}
+
+				WithdrawConsequence::Success
+			}
+			id if id == T::GetNativeCurrencyId::get() => {
+				<T::NativeCurrency as fungible::Inspect<_>>::can_withdraw(who, amount)
+			}
+			_ => <T::MultiCurrency as fungibles::Inspect<_>>::can_withdraw(asset_id, who, amount),
+		}
+	}
+}
+
+impl<T: Config> fungibles::Mutate<T::AccountId> for Pallet<T> {
+	fn mint_into(asset_id: Self::AssetId, who: &T::AccountId, amount: Self::Balance) -> DispatchResult {
+		if amount.is_zero() {
+			return Ok(());
+		}
+
+		match asset_id {
+			CurrencyId::Erc20(_) => return Err(Error::<T>::Erc20InvalidOperation.into()),
+			id if id == T::GetNativeCurrencyId::get() => {
+				<T::NativeCurrency as fungible::Mutate<_>>::mint_into(who, amount)
+			}
+			_ => <T::MultiCurrency as fungibles::Mutate<_>>::mint_into(asset_id, who, amount),
+		}?;
+		Self::deposit_event(Event::Deposited {
+			currency_id: asset_id,
+			who: who.clone(),
+			amount,
+		});
+		Ok(())
+	}
+
+	fn burn_from(
+		asset_id: Self::AssetId,
+		who: &T::AccountId,
+		amount: Self::Balance,
+	) -> Result<Self::Balance, DispatchError> {
+		if amount.is_zero() {
+			return Ok(amount);
+		}
+
+		let actual = match asset_id {
+			CurrencyId::Erc20(_) => return Err(Error::<T>::Erc20InvalidOperation.into()),
+			id if id == T::GetNativeCurrencyId::get() => {
+				<T::NativeCurrency as fungible::Mutate<_>>::burn_from(who, amount)
+			}
+			_ => <T::MultiCurrency as fungibles::Mutate<_>>::burn_from(asset_id, who, amount),
+		}?;
+		Self::deposit_event(Event::Withdrawn {
+			currency_id: asset_id,
+			who: who.clone(),
+			amount: actual,
+		});
+		Ok(actual)
+	}
+}
+
+impl<T: Config> fungibles::Transfer<T::AccountId> for Pallet<T> {
+	fn transfer(
+		asset_id: Self::AssetId,
+		source: &T::AccountId,
+		dest: &T::AccountId,
+		amount: Self::Balance,
+		keep_alive: bool,
+	) -> Result<Self::Balance, DispatchError> {
+		if amount.is_zero() || source == dest {
+			return Ok(amount);
+		}
+
+		let transferred_amount = match asset_id {
+			CurrencyId::Erc20(contract) => {
+				let sender = T::AddressMapping::get_evm_address(source).ok_or(Error::<T>::EvmAccountNotFound)?;
+				let origin = T::EVMBridge::get_origin().ok_or(Error::<T>::RealOriginNotFound)?;
+				let origin_address = T::AddressMapping::get_or_create_evm_address(&origin);
+				let address = T::AddressMapping::get_or_create_evm_address(dest);
+				T::EVMBridge::transfer(
+					InvokeContext {
+						contract,
+						sender,
+						origin: origin_address,
+					},
+					address,
+					amount,
+				)?;
+				Ok(amount)
+			}
```

Comment:
duplicated with `MultiCurrency` transfer impl https://github.com/AcalaNetwork/Acala/blob/80ba6122a8adb3ecff33154e0b3415ad55162db4/modules/currencies/src/lib.rs#L339-L353

Reply:
Looks like there's quite a few duplicates, I'm working through to clean it up